crates/solidity/testing/snapshots/cst_output/NumericExpression/hex_uppercase_prefix/generated/0.5.0-failure.yml

```diff
@@ -5,13 +5,15 @@ Source: >
 
 Errors: # 1 total
   - >
-    Error: Expected DecimalLiteral or HexLiteral.
-       ╭─[crates/solidity/testing/snapshots/cst_output/NumericExpression/hex_uppercase_prefix/input.sol:1:1]
+    Error: Expected DaysKeyword or EtherKeyword or FinneyKeyword or HoursKeyword or MinutesKeyword or SecondsKeyword or SzaboKeyword or WeeksKeyword or WeiKeyword.
+       ╭─[crates/solidity/testing/snapshots/cst_output/NumericExpression/hex_uppercase_prefix/input.sol:1:2]
        │
      1 │ 0X123456789
-       │ ─────┬─────  
-       │      ╰─────── Error occurred here.
+       │  ─────┬────  
+       │       ╰────── Error occurred here.
     ───╯
 
 Tree:
-  - SKIPPED (Token): "0X123456789" # 0..11
+  - NumericExpression (Rule): # 0..11 "0X123456789"
+      - DecimalLiteral (Token): "0" # 0..1
```

Comment:
Should we add "not followed by IdentifierPart" for numbers to prevent this? the `DecimalLiteral` is not valid (in any version) if followed by it.

Reply:
`0wei` is now legal, and it shouldn't be.

Reply:
Actually remix requires ! `IdentifierStart`. I don't think this was ever in our grammar? Since we removed the not-followed-by construct this isn't available. As an explicit test is in solc, we need to add it back. Not this PR. Created an issue for it.

Reply:
>  I don't think this was ever in our grammar?

The previous scanner was able to correctly reject `0wei`, and now it accepts it. This (#569 ?) would be a regression introduced by this PR, which would break `main` branch, blocking us from releasing this and any other fixes to users.

Not suggesting we add `not followed by` again if you already removed it (I'm glad it was removed from keywords), but if it was replaced by the new scanner architecture, then I suggest handling this case as well.

Reply:
I can't think of a solution other than `not_followed_by`. This is a strange requirement given that no ambiguity is introduced without it.

So with either release with this false positive, or we re-add not_followed_by. The alternative is a significant hack in the scanner generator to effectively add negative lookahead to a certain rule.

Identifiers are different - identifier vs. keyword is a generic issue in all languages. `DecimialLiteral` not followed by `IdentifierStart` is not.

This PR fixes a lot and adds this one regression. But if we feel that we must never regress, even with an aggregate inprovement, then I would re-add `not_followed_by`, which is the right thing to do.

My vote is for the aggregate improvement and another PR, especially given that this is a trivial regression with a non-trivial cause.

Reply:
And BTW I am actually happy for `not_followed_by` to be reintroduced because it is a general mechanism, rather than the identifier vs. keyword mechanism we have now. That mechanism (which I introduced) is, IMO, a mistake.

Reply:
> This PR fixes a lot and adds this one regression. But if we feel that we must never regress, even with an aggregate improvement, then I would re-add not_followed_by, which is the right thing to do.

I'm fine with partial work, or even a little duplication between PRs. For example, we now have two grammar definitions (Yaml and DSL) that we have to update/modify together moving forward, until I'm done with the procedural macros and AST types. But I hope to never introduce regressions/incorrectness, since we want to release the fixes in this PR (and the error recovery additions) as soon as it is merged.

In that case, I suggest bringing back `not followed by`. I just discovered one bug (in `NumericExpression`), but I'm sure there are more cases that we are missing. But if you think this is throw-away work. or there is anything else I'm missing, please feel free to push back.

Reply:
I'm worried about this PR lasting forever.

It's not throw away, even an FSM would want negative lookahead. Actually, I thought you were really keen on removing it :)

Reply:
F2f: let's merge as-is, and the first following PR should fix it correctly?

Reply:
yes, add non_followed_by next PR.